server/src/main/java/org/apache/druid/segment/metadata/AbstractSegmentMetadataCache.java

```diff
@@ -431,10 +431,7 @@ public Map<SegmentId, AvailableSegmentMetadata> getSegmentMetadataSnapshot()
   @Nullable
   public AvailableSegmentMetadata getAvailableSegmentMetadata(String datasource, SegmentId segmentId)
   {
-    if (!segmentMetadataInfo.containsKey(datasource)) {
-      return null;
-    }
-    return segmentMetadataInfo.get(datasource).get(segmentId);
+    return segmentMetadataInfo.getOrDefault(datasource, new ConcurrentSkipListMap<>()).get(segmentId);
```

Comment:
Why would you create a new object if its not used. Less GC that way. 
Isn't the older code more performant ?

Reply:
Removed creating a new object. 
In older there is a possibility of race since the datasource missing check is a separate statement.

Reply:
Reverted this change.

Reply:
My original branch was outdated, there was a race in the original implementation which was fixed in https://github.com/apache/druid/pull/16981.